CONTRIBUTING.md

```diff
@@ -3,104 +3,192 @@
-
-#### Branching & history
-
-- Use a feature branch, not master.
-- Rebase your feature branch onto (upstream) master before raising the PR.
-- Keep up to date with changes in (upstream) master so your PR is easy to merge.
-- Try to actively tidy your history: combine related commits with interactive
-  rebasing etc. If your PR is still `[WIP]` don't be afraid to force-push to
-  your feature branch to tidy your history.
-
-### For code PRs
+- Be descriptive in your pull request description: what is it for, why is it
+  needed, etc.
+- Do ***not*** make cosmetic changes to unrelated files in the same pull
+  request. This creates noise, making reviews harder to do. If your text
+  editor strips all trailing whitespace in a file when you edit it, disable
+  it.
+
+### Tagging in the issue tracker
+
+When submitting pull requests (commonly referred to as "PRs"), include one of
+the following tags prepended to the title:
+
+- `[WIP]` - Work In Progress. The PR will change, so while there is no
+  immediate need for review, the submitter still might appreciate it.
+- `[RFC]` - Request For Comment. The PR needs reviewing and/or comments.
+- `[RDY]` - The PR is ready to be merged. The PR has been reviewed by at least
+  one other person and has no outstanding issues.
+
+Assuming the above criteria has been met, feel free to change your PR's tag
+yourself, as opposed to waiting for a contributor to do it for you.
+
+### Branching & history
+
+- Do ***not*** work on your PR on the master branch, [use a feature branch
+  instead][git-feature-branch].
+- [Rebase your feature branch onto][git-rebasing] (upstream) master before
+  opening the PR.
+- Keep up to date with changes in (upstream) master so your PR is easy to
+  merge.
+- [Try to actively tidy your history][git-history-rewriting]: combine related
+  commits with interactive rebasing, split monolithic commits into multiple
+  ones, etc. If your PR is still `[WIP]` don't be afraid to force-push to your
+  feature branch to tidy your history.
+
+### For code pull requests
 
 #### Testing
 
-- We are unlikely to merge your PR if the Travis build fails.
-- The Travis build does not currently run the tests under valgrind, but you
-  are encouraged to do so locally.
+We are unlikely to merge your PR if the Travis build fails:
+
+- Travis builds are compiled with the [`-Werror`][gcc-warnings] flag, so
+  if your PR introduces any compiler warnings then the Travis build will fail.
+- If any unit tests fail, the Travis build will fail. They can be run
+  locally with `make test`, which is generally much faster than waiting on
+  Travis, so should be done prior to submitting your PR. Unit tests
+  passing locally doesn't guarantee they'll pass in your Travis build, as
+  multiple different builds using different platforms and compilers are
+  used to test your code.
+- The Travis build does not currently run the tests under
+  [valgrind][valgrind], but you're encouraged to do so locally.
```

Comment:
Travis does run Valgrind (GCC/Linux build). To do so locally, you can do something like `VALGRIND=1 make test`.